### PR TITLE
impl(common): add test for timer cancellation on Shutdown

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531)

### CODE REVIEW

File: google/cloud/internal/timer_queue.h [hunk 1]
```diff
@@ -33,16 +33,19 @@ class TimerQueue {
   TimerQueue(TimerQueue const&) = delete;
   TimerQueue& operator=(TimerQueue const&) = delete;
 
-  // Adds a timer to the queue.
+  // Adds a timer to the queue. If the `Shutdown` has already been called,
+  // calls to `Schedule` return immediately with a `StatusCode::kCancelled`.
```
Comment (devbww) [inline google/cloud/internal/timer_queue.h:37]: It returns a future that is ready with a cancelled status.

File: google/cloud/internal/timer_queue.h [hunk 2]
```diff
@@ -33,16 +33,19 @@ class TimerQueue {
...
   future<StatusOr<std::chrono::system_clock::time_point>> Schedule(
       std::chrono::system_clock::time_point tp);
 
-  // Signals all threads that have called Service to return. Does not modify
-  // remaining timers. If desired, CancelAll can be called.
+  // Signals all threads that have called `Service` to return. Any subsequent
+  // calls to `Schedule` are refused. Additionally, all outstanding timers are
```
Comment (devbww) [inline google/cloud/internal/timer_queue.h:42]: "refused" is less than clear.

File: google/cloud/internal/timer_queue_test.cc [hunk 3]
```diff
@@ -81,6 +81,23 @@ TEST(TimerQueueTest, ScheduleEarlierTimerSingleRunner) {
...
+  auto now = std::chrono::system_clock::now();
+  std::thread t([&tq] { tq.Service(); });
+  auto f = tq.Schedule(now + duration);
+  auto f2 = tq.Schedule(now + duration);
+
+  tq.Shutdown();
+  t.join();
+
+  auto expire_time = f.get();
+  EXPECT_THAT(expire_time, StatusIs(StatusCode::kCancelled));
```
Comment (devbww) [inline google/cloud/internal/timer_queue_test.cc:96]: These are still timing dependent, albeit with admitted low probability, but I don't have a better solution short of checking that they are either cancelled or yield a time in the past.

File: google/cloud/internal/timer_queue_test.cc [hunk 4]
```diff
@@ -161,6 +178,27 @@ TEST(TimerQueueTest, ScheduleEarlierTimerMultipleRunner) {
   EXPECT_GT(*later_expire_time, *earlier_expire_time);
 }
 
+TEST(TimerQueueTest, ShutdownWithPendingTimersCancelledMultipleRunner) {
```
Comment (devbww) [inline google/cloud/internal/timer_queue_test.cc:181]: Perhaps these "Single/Multiple" tests should have their common-ness extracted by making them tests parameterized on thread count.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/internal/timer_queue.h` line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531#-NLO34FIINnwE7bi83Eu-r1-37:-NLRmrKgC4RnP0903Y6e:bat1m5k) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12c1b222ef05eb501079e1fbed10d1e7b6c9b0c8/google/cloud/internal/timer_queue.h#L37)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It returns a future that is ready with a cancelled status.
</blockquote></details>

The return type is indicated on the next line, I didn't think it was worth duplicating the type, just describing what the value would be.
___
*[`google/cloud/internal/timer_queue.h` line 42 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531#-NLO34FIINnwE7bi83Eu-r1-42:-NLRn-qo9LCjKBJgWcSj:b4oedu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12c1b222ef05eb501079e1fbed10d1e7b6c9b0c8/google/cloud/internal/timer_queue.h#L42)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

"refused" is less than clear.
</blockquote></details>

Removed unclear statement as it duplicated comments on `Schedule`.
___
*[`google/cloud/internal/timer_queue_test.cc` line 96 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531#-NLO34FIINnwE7bi83Ev-r1-96:-NLRng0k6wFK5qlEsP8q:b-17axp2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12c1b222ef05eb501079e1fbed10d1e7b6c9b0c8/google/cloud/internal/timer_queue_test.cc#L96)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

These are still timing dependent, albeit with admitted low probability, but I don't have a better solution short of checking that they are either cancelled or yield a time in the past.
</blockquote></details>

Moved starting the runners until after the Schedule and Shutdown calls to try and avoid timing issues.
___
*[`google/cloud/internal/timer_queue_test.cc` line 181 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10531#-NLO34FIINnwE7bi83Ev-r1-181:-NLRnlvR8bO1NzW5jVYx:b-1smy4f) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12c1b222ef05eb501079e1fbed10d1e7b6c9b0c8/google/cloud/internal/timer_queue_test.cc#L181)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Perhaps these "Single/Multiple" tests should have their common-ness extracted by making them tests parameterized on thread count.
</blockquote></details>

Perhaps, but there's also competing style directives on keeping test code "local/readable" by not abstracting commonality into fixtures/helper functions. If someone else feels the need to parameterize these tests I won't object.

**Pull request closed without merging**